### PR TITLE
Extractor and evm entities tests

### PR DESCRIPTION
**Pull request by kayibal**
Adding unit tests for evm entities and extractor.

### CODE REVIEW

File: src/extractor/evm/ambient.rs [hunk 1]
```diff
@@ -270,16 +272,20 @@ where
 
         let msg = match evm::BlockStateChanges::try_from_message(raw_msg, &self.name, self.chain) {
             Ok(changes) => changes,
-            Err(ExtractionError::Empty) => return Ok(None),
+            Err(ExtractionError::Empty) => {
+                return {
+                    self.update_cursor(inp.cursor).await;
```
Comment (zizou0x) [inline src/extractor/evm/ambient.rs]: Why is this line in the return block? Does it change anything?

File: src/extractor/evm/ambient.rs [hunk 2]
```diff
@@ -312,3 +318,121 @@ where
...
+    #[tokio::test]
+    async fn test_get_cursor() {
+        let mut gw = MockAmbientGateway::new();
+        gw.expect_get_cursor()
+            .withf(|name, chain| name == "vm:ambient" && chain == &Chain::Ethereum)
+            .times(1)
+            .returning(|_, _| Ok("cursor".into()));
+        let extractor = AmbientContractExtractor::new("vm:ambient", Chain::Ethereum, gw)
+            .await
+            .expect("extractor init ok");
```
Comment (pistomat) [inline src/extractor/evm/ambient.rs:339]: Note: I found the "expect as error message" a bit confusing, but then I read https://doc.rust-lang.org/std/error/index.html#common-message-styles and now the "expect as precondition" makes sense to me now. I should update my expect messages in the ws tests.

Comment (kayibal) [reply src/extractor/evm/ambient.rs:339]: This little paragraph made me get it:

> Hint: If you’re having trouble remembering how to phrase expect-as-precondition style error messages remember to focus on the word “should” as in “env variable should be set by blah” or “the given binary should be available and executable by the current user”.

File: src/extractor/evm/ambient.rs (hunk 1 above)
Comment (kayibal) [reply src/extractor/evm/ambient.rs]: Mh no I think I should fix that.

File: src/extractor/evm/ambient.rs [hunk 3]
```diff
@@ -312,3 +326,121 @@ where
...
+        let exp = Ok(Some(()));
+
+        let res = extractor
+            .handle_tick_scoped_data(inp)
+            .await
+            .map(|o| o.map(|_| ()));
+
+        assert_eq!(res, exp);
+        assert_eq!(extractor.get_cursor().await, "cursor@420");
+    }
```
Comment (carloszanella) [inline src/extractor/evm/ambient.rs:380]: can't review the tests very well as I am unfamiliar with the extractor logic

File: src/extractor/evm/utils.rs [hunk 4]
```diff
@@ -6,17 +6,31 @@ pub fn u256_to_bytes(v: &U256) -> Vec<u8> {
     bytes32.to_vec()
```
Comment (pistomat) [inline src/extractor/evm/utils.rs]: Out of scope: Is this an alias for [encode()](https://docs.rs/ethers/latest/ethers/core/types/struct.U256.html#method.encode-1)?

File: src/extractor/evm/utils.rs [hunk 5]
```diff
@@ -6,17 +6,31 @@ pub fn u256_to_bytes(v: &U256) -> Vec<u8> {
...
-///
-/// The db id is required to provide additional error context in case the
-/// parsing fails.
+/// Parses an evm address hash from a slice
 pub fn parse_id_h160(v: &[u8]) -> Result<H160, String> {
     if v.len() != 20 {
-        return Err(format!("Invalid contract address found: {}", hex::encode(v)))
+        return Err(format!("H160: Invalid address format: 0x{}", hex::encode(v)))
     }
     Ok(H160::from_slice(v))
```
Comment (pistomat) [inline src/extractor/evm/utils.rs]: You can just use `.map_err()`, because wrong length is the only way it panics https://docs.rs/ethers/latest/ethers/core/types/struct.H160.html#panics-1

File: src/extractor/evm/utils.rs [hunk 6]
```diff
@@ -6,17 +6,31 @@ pub fn u256_to_bytes(v: &U256) -> Vec<u8> {
...
 }
 
+// Parse an address potentially zero left padding the value
+pub fn pad_and_parse_h160(v: &[u8]) -> Result<H160, String> {
```
Comment (pistomat) [inline src/extractor/evm/utils.rs:53]: Why are we ever using padding? Do we omit leading zeros somewhere for the sake of storage efficiency?

As H160, Address and all related types are fixed length, I would consider any input with length other than 20 an error and try to fix it in the source.

File: src/extractor/evm/utils.rs [hunk 7]
```diff
@@ -45,3 +59,73 @@ pub fn parse_u256_slot_entry(
     let k = U256::from_big_endian(raw_key);
     Ok((k, v))
 }
+
+// Parse a 32 bytes value potentially zero left padding the value
+pub fn pad_and_parse_32bytes<T>(v: &[u8]) -> Result<T, String>
```
Comment (pistomat) [inline src/extractor/evm/utils.rs:84]: Again, I think padding any fixed size types is a mistake and should be resolved at the source of data.

File: src/extractor/evm/mod.rs [hunk 8]
```diff
@@ -93,12 +100,21 @@ impl Account {
...
+    /// missing, it will use the corresponding types default.
+    /// Will use the associated transaction as creation, balance and code modify
+    /// transaction.
     fn from(value: &AccountUpdateWithTx) -> Self {
         let empty_hash = H256::from(keccak256(Vec::new()));
         Account::new(
             value.chain,
             value.address,
-            format!("{:x}", value.address),
+            format!("0x{:x}", value.address),
```
Comment (pistomat) [inline src/extractor/evm/mod.rs]: I would prefer using [`encode_hex()`](https://docs.rs/ethers/latest/ethers/core/types/struct.H160.html?#method.encode_hex)

File: src/extractor/evm/mod.rs [hunk 9]
```diff
@@ -137,12 +153,35 @@ impl AccountUpdate {
...
+    ///
+    /// * `other`: An instance of `AccountUpdate`. The attribute values and keys
+    /// of `other` will overwrite those of `self`.
     fn merge(&mut self, other: AccountUpdate) -> Result<(), ExtractionError> {
         if self.address != other.address {
-            return Err(ExtractionError::Unknown(
-                "Can't merge AccountUpdates from differing identities".into(),
-            ))
+            return Err(ExtractionError::Unknown(format!(
+                "Can't merge AccountUpdates from differing identities; Expected 0x{:x}, got 0x{:x}",
```
Comment (pistomat) [inline src/extractor/evm/mod.rs]: Insead of prefixing `0x` manually, you can use the [LowerHex implementation ](https://stackoverflow.com/a/75509511)

File: src/extractor/evm/utils.rs (hunk 4 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs]: Oh I didn't know this existed... well would probably be better to use `encode()` here..

File: src/extractor/evm/utils.rs (hunk 5 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs]: But it panics so there is no err, right?

File: src/extractor/evm/utils.rs (hunk 6 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs:53]: I think I need these for protobuf... I think this is probably how we get the data from streamingfasts block storage..

File: src/extractor/evm/utils.rs (hunk 7 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs:84]: Will check where these are coming from and if I have control over it fix them, although I think we don't have control over it. 
So I can either do the padding here .. or within the substreams handlers. To reduce bandwith I would probably prefer to do it here.

File: src/extractor/evm/utils.rs (hunk 5 above)
Comment (pistomat) [reply src/extractor/evm/utils.rs]: You are right, we should use `.decode()`, which returns an error

File: src/extractor/evm/utils.rs (hunk 7 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs:84]: So I spent some time looking into it and yes these padding methods are used when reading from streamingfasts protobuf message. I think they do not store leading zeros but this is not so easy to find our right away. (I would need to write a spkg to get the blocks - this is useful anyway for testing so I'll do it sometime soon).
TLDR: I'll leave it for now but I agree that it would be preferrable to have as little ambiguity as possible and streamingfast should probably be changed so it has the correct fixed size in case it hasn't.

File: src/extractor/evm/utils.rs (hunk 5 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs]: ~~I went looking, couldn't find a `decode` method on H160, nor on &[u8]. But yeah couldn't find a non panicking way to convert a variable length slice to ethers types.~~

EDIT: Found it

File: src/extractor/evm/mod.rs (hunk 8 above)
Comment (kayibal) [reply src/extractor/evm/mod.rs]: If I use encode hex I get `"0x000000000000000000000000e688b84b23f322a994a53dbf8e15fa82cdb71127"`

File: src/extractor/evm/utils.rs (hunk 4 above)
Comment (kayibal) [reply src/extractor/evm/utils.rs]: Well turns our the abi encode and decode methods are probably not a good idea to use unless you are doing abi encoding decoding - because they assume everything to be 32 bytes including H160.
